pkg/ebpf/filters.go

```diff
@@ -37,3 +41,318 @@ type Filter struct {
 	BinaryFilter      *filters.BPFBinaryFilter
 	Follow            bool
 }
+
+func NewFilterScope() *FilterScope {
+	return &FilterScope{
+		ID:                0,
+		EventsToTrace:     map[events.ID]string{},
+		UIDFilter:         filters.NewBPFUInt32Filter(UIDFilterMap),
+		PIDFilter:         filters.NewBPFUInt32Filter(PIDFilterMap),
+		NewPidFilter:      filters.NewBoolFilter(),
+		MntNSFilter:       filters.NewBPFUIntFilter(MntNSFilterMap),
+		PidNSFilter:       filters.NewBPFUIntFilter(PidNSFilterMap),
+		UTSFilter:         filters.NewBPFStringFilter(UTSFilterMap),
+		CommFilter:        filters.NewBPFStringFilter(CommFilterMap),
+		ContFilter:        filters.NewBoolFilter(),
+		NewContFilter:     filters.NewBoolFilter(),
+		ContIDFilter:      filters.NewContainerFilter(CgroupIdFilterMap),
+		RetFilter:         filters.NewRetFilter(),
+		ArgFilter:         filters.NewArgFilter(),
+		ContextFilter:     filters.NewContextFilter(),
+		ProcessTreeFilter: filters.NewProcessTreeFilter(ProcessTreeFilterMap),
+		BinaryFilter:      filters.NewBPFBinaryFilter(BinaryFilterMap),
+		Follow:            false,
+	}
+}
+
+const MaxFilterScopes = 64
+
+func isIDInRange(id int) bool {
+	return id >= 0 && id < MaxFilterScopes
+}
+
+func (fs FilterScope) IsEventTraceable(event events.ID) bool {
+	if _, found := fs.EventsToTrace[event]; found {
+		return true
+	}
+
+	return false
+}
+
+// TODO: add locking mechanism as scopes will change at runtime
+type FilterScopes struct {
+	filterScopesArray        [MaxFilterScopes]*FilterScope // underlying filter scopes array
+	filterScopesMap          map[*FilterScope]int          // stores only enabled scopes
```

Comment:
> it prevents the same scope object (as in the actual object in memory, since this is a pointer), being used in two separate places.

Indeed. That's the point of our use here, two maps and one array, all linked through scope pointers.

> As such I could construct the same exact `FilterScope` in different pointers and it should work fine (try it with `-t 4:comm=ls -t 5:comm=ls`).

Yes, they are the same in logic. As we didn't draw the next step completely, I preferred to leave it as it is, a different object.

> I assume you are aware of this but it does make the AlreadySet error a bit unnecessary, and should be probably documented for clarity sake.

We still don't know if this type of error should be disregarded. It will depend on the type of API we want to provide for, let's say, a management web interface. Could this interface to pick scopes and reuse them? Would it be clever to reuse the object or just rebuild it? Open to discuss.